src/VisualStudio/Next/Remote/ServiceHubRemoteHostClient.JsonRpcSnapshot.cs

```diff
@@ -0,0 +1,159 @@
+            }
+
+            public override Task<T> InvokeAsync<T>(string targetName, params object[] arguments)
+            {
+                CancellationToken.ThrowIfCancellationRequested();
+
+                return _serviceClient.InvokeAsync<T>(targetName, arguments.Concat(ChecksumScope.SolutionChecksum.Checksum.ToArray()).ToArray());
+            }
+
+            public override Task InvokeAsync(string targetName, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task> funcWithDirectStreamAsync)
+            {
+                CancellationToken.ThrowIfCancellationRequested();
+
+                return _serviceClient.InvokeAsync(targetName, arguments.Concat(ChecksumScope.SolutionChecksum.Checksum.ToArray()).ToArray(), funcWithDirectStreamAsync, CancellationToken);
+            }
+
+            public override Task<T> InvokeAsync<T>(string targetName, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task<T>> funcWithDirectStreamAsync)
+            {
+                CancellationToken.ThrowIfCancellationRequested();
+
+                return _serviceClient.InvokeAsync<T>(targetName, arguments.Concat(ChecksumScope.SolutionChecksum.Checksum.ToArray()).ToArray(), funcWithDirectStreamAsync, CancellationToken);
+            }
+
+            protected override void OnDisposed()
+            {
+                // dispose cancellation registration
+                _cancellationRegistration.Dispose();
+
+                // dispose service and snapshot channels
+                _serviceClient.Dispose();
+                _snapshotClient.Dispose();
+            }
+
+            private class ServiceJsonRpcClient : JsonRpcClient
+            {
+                private readonly object _callbackTarget;
+
+                public ServiceJsonRpcClient(Stream stream, object callbackTarget) : base(stream)
+                {
+                    // this one doesn't need cancellation token since it has nothing to cancel
+                    _callbackTarget = callbackTarget;
+                }
+
+                protected override object GetTarget() => _callbackTarget;
+            }
+
+            private class SnapshotJsonRpcClient : JsonRpcClient
+            {
+                private readonly JsonRpcSession _owner;
+                private readonly CancellationTokenSource _source;
+
+                public SnapshotJsonRpcClient(JsonRpcSession owner, Stream stream) :
+                    base(stream)
+                {
+                    _owner = owner;
+                    _source = new CancellationTokenSource();
+                }
+
+                private ChecksumScope ChecksumScope => _owner.ChecksumScope;
+                protected override object GetTarget() => this;
+
+                public async Task RequestAssetAsync(int serviceId, byte[] checksum, string streamName)
+                {
```

Comment:
right now, logging. with env var set, service hub will log this information so I can track communication.